tools/codegen/gen.py

```diff
@@ -259,38 +261,29 @@ def __call__(self, f: NativeFunction) -> Optional[str]:
 
                 has_tensor_options = any(isinstance(a.argument, TensorOptionsArguments) for a in args)
 
-                if local.use_c10_dispatcher() == UseC10Dispatcher.full:
-                    cuda_guard_from_tensor_options = """\
-    const DeviceGuard device_guard(device_or_default(device));
-"""
-                else:
-                    assert local.use_c10_dispatcher() in [UseC10Dispatcher.with_codegenerated_unboxing_wrapper,
-                                                          UseC10Dispatcher.hacky_wrapper_for_legacy_signatures]
-                    cuda_guard_from_tensor_options = """\
-    const DeviceGuard device_guard(options.device());
-"""
-
-                # TODO: There is probably a simpler version of this that
-                # works just as well.
-                if f.device_guard and is_generic_dispatch_key(self.dispatch_key) and has_tensor_options:
-                    cuda_guard = cuda_guard_from_tensor_options
-                elif f.device_guard and is_cuda_dispatch_key(self.dispatch_key) and has_tensor_options:
-                    cuda_guard = f"""\
-    globalContext().lazyInitCUDA();
-    {cuda_guard_from_tensor_options}
-"""
+                # Initialization happens on construction of CUDA tensors, so
+                # that all methods/functions that operates on CUDA tensors can
+                # assume initialization has already occurred
+                if is_cuda_dispatch_key(self.dispatch_key) and has_tensor_options:
+                    init_cuda = "globalContext().lazyInitCUDA();"
+
+                if f.device_guard and has_tensor_options:
+                    if local.use_c10_dispatcher() is UseC10Dispatcher.full:
+                        cuda_guard = "const DeviceGuard device_guard(device_or_default(device));"
+                    else:
+                        assert local.use_c10_dispatcher() in [UseC10Dispatcher.with_codegenerated_unboxing_wrapper,
+                                                              UseC10Dispatcher.hacky_wrapper_for_legacy_signatures]
+                        cuda_guard = "const DeviceGuard device_guard(options.device());"
                 elif f.device_guard and device_of is not None:
-                    cuda_guard = f"""\
-    const OptionalDeviceGuard device_guard(device_of({device_of}));
-"""
+                    cuda_guard = f"const OptionalDeviceGuard device_guard(device_of({device_of}));"
                 else:
-                    cuda_guard = """\
-    // DeviceGuard omitted
-"""
+                    cuda_guard = "// DeviceGuard omitted"
```

Comment:
High-level comment on how this logic is structured: it might be clarifying to use `has_tensor_options` as the outer partition - I think it cleans up dependencies and  better aligns with the semantics of the wrapper we're generating (note: the snippet is big bc it also tucks `self_args`, `candidate_args` and `device_of` in the `else` where they belong):
```
device_guard = "// DeviceGuard omitted" # default

if has_tensor_options:
  # kernel is creating a tensor
  if is_cuda_dispatch_key(self.dispatch_key):
    # initialize CUDA on construction of CUDA tensors
    init_cuda = "globalContext().lazyInitCUDA();"
  if f.device_guard:
    if local.use_c10_dispatcher() is UseC10Dispatcher.full:
      device_guard = "const DeviceGuard device_guard(device_or_default(device));"
    else:
      device_guard = "const DeviceGuard device_guard(options.device());"
else:
  # kernel is operating on existing tensors
  if f.device_guard:
    self_args = (a for a in f.func.arguments if a.name == "self")
    # There is precedence for which argument we use to do
    # device guard.  This describes the precedence order.
    candidate_args = itertools.chain(self_args, f.func.out_arguments, f.func.arguments)
    # Only tensor like arguments are eligible
    device_of = next((f'{a.name}' for a in candidate_args if a.type.is_tensor_like()), None)
    if device_of is not None:
      device_guard = f"const OptionalDeviceGuard device_guard(device_of({device_of}));"

```